coding-questions-python/longest-palindromic-substring.py

```diff
@@ -0,0 +1,74 @@
+'''
+Given a string, find the longest substring which is palindrome. For example, if the given string is “forgeeksskeegfor”, the output should be “geeksskeeg”.
+
+Suggested approach:
+Dynamic programming -- maintain a 2D array that stores a boolean at each index
+	- if inex i,j is true, that means that the substring from index i to j (inclusive) is a palindrome
+This way, we can use pre-computed results of smaller lengths to find the current result for the (current) lengths
+'''
+
+
+def init():
+	print(dummy)
+
+def longestPalindromicSubstring(str):
+	cache = []
+	longestLength = -1
+	longestSubstring = ""
+
+	
+	#init 2D array with False at all indices
+	for i in range(len(str)):
+		cache.append([])
+		for j in range(len(str)):
+			cache[i].append(False)
+
+	#set the values for substrings of length 1 and 2
+	for i in range(len(cache)):
+		cache[i][i] = True
+		longestSubstring = setSubstring(str, i, i, longestSubstring, longestLength)
```

Comment:
`longestSubstring` is being copied each time a longer substring is seen. You might want to instead cache the start and end indices of the longestSubstring.

Reply:
Can you please clarify? Even if I just store the indices, the helper function `setSubstring()` will still need a "copy" of the str right? The only way to avoid this copying is to perform that operation in line rather than using this helper function. This also means that the code will be duplicated, not once but twice.